presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/StochasticPriorityQueue.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.resourceGroups;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+final class StochasticPriorityQueue<E>
+        implements UpdateablePriorityQueue<E>
+{
+    private final Map<E, Node<E>> index = new HashMap<>();
+
+    // This is a Fenwick tree, where each node has weight equal to the sum of its weight
+    // and all its children's weights
+    private Node<E> root;
+
+    @Override
+    public boolean addOrUpdate(E element, int priority)
+    {
+        checkArgument(priority > 0, "priority must be positive");
+        if (root == null) {
+            root = new Node<>(Optional.empty(), element);
+            root.setTickets(priority);
+            index.put(element, root);
+            return true;
+        }
+        Node<E> node = index.get(element);
+        if (node != null) {
+            node.setTickets(priority);
+            return false;
+        }
+
+        node = root.addNode(element, priority);
+        index.put(element, node);
+        return true;
+    }
+
+    @Override
+    public boolean contains(E element)
+    {
+        return index.containsKey(element);
+    }
+
+    @Override
+    public boolean remove(E element)
+    {
+        Node<E> node = index.remove(element);
+        if (node != null) {
+            if (node.isLeaf() && node.equals(root)) {
+                root = null;
+            }
+            else if (node.isLeaf()) {
+                node.remove();
+            }
+            else {
+                // Remove a leaf and swap values with it
```

Comment:
```java
// This is an intermediate node.  Instead of removing it directly, remove a leaf 
// and then replace the data in this node with the data in the leaf.  This way
// we don't have to reorganize the tree structure.